_posts/2020-03-29-all-you-need-is-lambda-1-booleans.md

```diff
@@ -0,0 +1,260 @@
+date:   2020-03-29 20:17:26
+categories: ["lambda calculus", "plt"]
+---
+
+# All you need is λ, part one: booleans
+
+Nearly a hundred years ago, Alonzo Church invented computation in the form of the simple, elegant, and utterly inscrutable lambda calculus. Along with Alan Turing, he then proved that anything computable can be computed in the lambda calculus. However, nearly as soon as we had digital computers, we started inventing programming languages, and with them a vast treasure of features, beautiful and terrible, many of which seem very hard to relate to the fundamental nature of computability, let alone the lambda calculus.
+
+While it’s true that anything which can be computed, period, can be computed in the lambda calculus, you might not want to: it’s austere, to say the least, and was not designed with modern sensibilities regarding readability in mind. We developed all those languages and features for a reason! Still, Church demonstrated not just that it was possible to compute anything computable with the lambda calculus, but also _how_ one might do so.
+
+
+## Unconditional λ
+
+The lambda calculus’ austerity is extreme: you don’t even have booleans. All you have are:
+
+1. Single-argument functions (“lambda abstractions,” but “lambdas,” “abstractions,” and “functions” will also be used interchangeably).
+
+    These are written `λ x . y`, for variable `x` and expression `y`, where `x` is now available as a bound variable in the body, and any enclosing definition of `x` is shadowed. (We shall assume strictly lexical scoping for the time being.)
+
+    In Haskell, we would write `\ x -> y` instead; in JavaScript, `function (x) { return y }` or `(x) => y`.
+
+2. Function application (“lambda application,” or “application” or “function calls”).
+
+    Applications are written as `x y`, for expressions x and y, and left-associated, i.e. `a b c` = `(a b) c` ≠ `a (b c)`. Function application binds tighter than lambda abstraction, i.e. `λ x . λ y . y x` = `λ x . λ y . (y x)` ≠ `λ x . (λ y . y) x`.
+
+    The syntax is the same in Haskell; in JavaScript, we would write `x(y)` or `a(b, c)`. Note however that since lambda calculus functions are all single-argument functions, a more direct (tho less idiomatic) equivalent for the latter would be `a(b)(c)`.
+
+3. Variables (written as names; bound by enclosing lambdas; no `let` bindings, no globals; we will feel free to give variables names both great and small, and will include non-alphanumeric characters in names as we see fit, since the paucity of syntax means there’s little risk of ambiguity).
+
+Since a lambda is the only way to introduce a value—it’s the only “literal” syntax, and there are no predefined global definitions or standard library—we can also see that the only kinds of runtime values must be closures, perhaps consisting of the name of the introduced variable, body of the lambda, & values of any variables it closed over when made (again, we assume strict lexical scoping).
```

Comment:
This sentence is long and a little hard to read. Perhaps consider splitting it into two sentences:

```suggestion
Since a lambda is the only way to introduce a value—it’s the only “literal” syntax, and there are no predefined global definitions or standard library. We can also see that the only kinds of runtime values must be closures, perhaps consisting of the name of the introduced variable, body of the lambda, and values of any variables it closed over when made (again, we assume strict lexical scoping).
```